File: src/storage.ts
import { createHash, randomBytes } from 'node:crypto';
import { closeSync, constants, existsSync, fstatSync, fsyncSync, lstatSync, mkdirSync, openSync, readFileSync, readdirSync, renameSync, rmSync, writeFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import ts from 'typescript';
import type { Run, RunSummary, Source } from '../web/src/types.ts';
export type { Run, RunSummary, Source } from '../web/src/types.ts';
export const projectRoot = resolve(import.meta.dirname, '..');
export const runIDPattern = /^[a-f0-9]{32}$/;
export const eventIDPattern = /^e[0-9]+$/;
export const randomID = () => randomBytes(16).toString('hex');
export const fileHash = (raw: string | Buffer) => createHash('sha256').update(raw).digest('hex');
export function object(value: unknown): Record<string, unknown> {
    if (!value || typeof value !== 'object' || Array.isArray(value))
        throw new Error('JSON object required');
    return value as Record<string, unknown>;
}
export function integerArgument(args: Record<string, unknown>, key: string, fallback: number): number {
    if (!Object.hasOwn(args, key))
        return fallback;
    if (!Number.isSafeInteger(args[key]))
        throw new Error('integer required');
    return args[key] as number;
}
export function text(value: unknown): string {
    if (typeof value !== 'string' || !value.isWellFormed())
        throw new Error('string required');
    return value;
}
export function ordinaryDirectory(path: string): void {
    const info = lstatSync(path);
    if (!info.isDirectory() || info.isSymbolicLink())
        throw new Error('ordinary directory required');
}
export function readOrdinaryFile(path: string, maxBytes = 64 * 1024 * 1024): Buffer {
    const fd = openSync(path, constants.O_RDONLY | constants.O_NOFOLLOW | constants.O_NONBLOCK);
    try {
        const info = fstatSync(fd);
        if (!info.isFile() || info.size > maxBytes)
            throw new Error('ordinary bounded file required');
        return readFileSync(fd);
    }
    finally {
        closeSync(fd);
    }
}
export function appendJSON(path: string, value: unknown): void {
    mkdirSync(dirname(path), { recursive: true, mode: 0o700 });
    ordinaryDirectory(dirname(path));
    const fd = openSync(path, constants.O_CREAT | constants.O_APPEND | constants.O_WRONLY | constants.O_NOFOLLOW, 0o600);
    try {
        if (!fstatSync(fd).isFile())
            throw new Error('ordinary file required');
        writeFileSync(fd, JSON.stringify(value) + '\n');
        fsyncSync(fd);
    }
    finally {
        closeSync(fd);
    }
}
export function atomicWrite(path: string, content: string, mode = 0o600): void {
    ordinaryDirectory(dirname(path));
    const temporary = join(dirname(path), `.loop-${randomID()}.tmp`);
    const fd = openSync(temporary, 'wx', mode);
    try {
        try {
            writeFileSync(fd, content);
            fsyncSync(fd);
        }
        finally {
            closeSync(fd);
        }
        renameSync(temporary, path);
    }
    finally {
        rmSync(temporary, { force: true });
    }
}
export function saveRun(path: string, run: Run): void { atomicWrite(path, JSON.stringify(run, null, 2) + '\n'); }
export function summarizeRun(run: Run): RunSummary {
    const { id, task, model, status, created_at, exercise, parent_run_id, conversation_id, conversation_turn } = run;
    return { id, task, model, status, created_at, ...(exercise ? { exercise } : {}), ...(parent_run_id ? { parent_run_id, conversation_id } : {}), conversation_turn };
}
export function sortedSummaries(history: Map<string, RunSummary>): RunSummary[] {
    return [...history.values()].sort((a, b) => b.created_at - a.created_at || a.id.localeCompare(b.id));
}
// Capture the exact source and assets once at server startup, before accepting runs.
export function sourceRecords(): {
    sources: Record<string, Source>;
    buildID: string;
} {
    const definitions: [
        string,
        string,
        string
    ][] = [
        ['context', 'context.ts', 'BuildTurnMessages'], ['loop', 'agent.ts', 'RunLoop'],
        ['coding_dispatch', 'coding.ts', 'ExecuteCoding'], ['command', 'coding.ts', 'RunExerciseTests'], ['write', 'coding.ts', 'WriteExerciseFile'],
        ['dispatch', 'tools.ts', 'ExecuteReadonly'], ['read', 'tools.ts', 'ReadFile'], ['list', 'tools.ts', 'ListFiles'],
    ];
    const sources: Record<string, Source> = {};
    const paths = ['index.html', 'tools.json', 'coding-tools.json', 'package.json',
        ...readdirSync(join(projectRoot, 'src')).filter(p => p.endsWith('.ts')).map(p => `src/${p}`),
        ...readdirSync(join(projectRoot, 'web/dist')).filter(p => p.endsWith('.js')).map(p => `web/dist/${p}`),
        ...readdirSync(join(projectRoot, 'testdata/ts-average')).map(p => `testdata/ts-average/${p}`)];
    const hash = createHash('sha256'), rawFiles = new Map<string, string>();
    for (const path of paths.sort()) {
        const raw = readFileSync(join(projectRoot, path), 'utf8');
        rawFiles.set(path, raw);
        hash.update(path).update('\0').update(raw);
    }
    for (const [key, file, name] of definitions) {
        const path = `src/${file}`, raw = rawFiles.get(path)!;
        const tree = ts.createSourceFile(path, raw, ts.ScriptTarget.Latest, true);
        const fn = tree.statements.find(node => ts.isFunctionDeclaration(node) && node.name?.text === name);
        if (!fn)
            throw new Error(`missing source function ${name}`);
        const start = fn.getStart(tree);
        sources[key] = { path, line: tree.getLineAndCharacterOfPosition(start).line + 1, code: raw.slice(start, fn.end) };
    }
    return { sources, buildID: hash.digest('hex') };
}
export function decodeHistory(raw: Buffer | string, id: string): Run {
    const run = object(JSON.parse(raw.toString()));
    if (!runIDPattern.test(id) || run.id !== id || run.engine !== 'typescript')
        throw new Error('invalid run identity');
    for (const key of ['task', 'model', 'workspace', 'answer', 'task_result', 'build_id'])
        text(run[key]);
    if (run.exercise !== undefined && run.exercise !== 'ts-average')
        throw new Error('invalid exercise');
    if (!['completed', 'failed', 'budget_exhausted'].includes(text(run.status)))
        throw new Error('unfinished run');
    for (const key of ['created_at', 'duration', 'model_requests', 'tool_calls', 'tool_errors']) {
        if (typeof run[key] !== 'number' || !Number.isFinite(run[key]) || run[key] < 0)
            throw new Error('invalid counters');
    }
    for (const key of ['model_requests', 'tool_calls', 'tool_errors', 'conversation_turn', 'max_requests']) {
        if (!Number.isSafeInteger(run[key]))
            throw new Error('invalid integer');
    }
    if ((run.max_requests as number) < 1 || (run.max_requests as number) > 8)
        throw new Error('invalid budget');
    if (run.error !== null)
        object(run.error);
    if (run.parent_run_id !== undefined) {
        if (!runIDPattern.test(text(run.parent_run_id)) || run.parent_run_id === id || !runIDPattern.test(text(run.conversation_id)) || run.conversation_id === id || (run.conversation_turn as number) < 2)
            throw new Error('invalid conversation link');
    }
    else if (run.conversation_id !== undefined || run.conversation_turn !== 1)
        throw new Error('invalid root');
    if (!Array.isArray(run.events) || !run.events.length)
        throw new Error('invalid events');
    const sources = object(run.source), ids = new Set<string>();
    for (const item of run.events) {
        const event = object(item), eventID = text(event.id);
        if (!eventIDPattern.test(eventID) || ids.has(eventID))
            throw new Error('invalid event identity');
        ids.add(eventID);
        for (const key of ['title', 'code', 'explanation'])
            text(event[key]);
        if (!['input', 'model', 'tool', 'control'].includes(text(event.kind)) || !['succeeded', 'failed'].includes(text(event.status)))
            throw new Error('invalid event state');
        for (const key of ['t', 'd', 'turn'])
            if (typeof event[key] !== 'number' || !Number.isFinite(event[key]) || event[key] < 0)
                throw new Error('invalid event timing');
        if (!Number.isSafeInteger(event.turn))
            throw new Error('invalid event turn');
        object(event.input);
        object(event.output);
        const source = object(sources[text(event.code)]);
        text(source.path);
        text(source.code);
        if (!Number.isSafeInteger(source.line) || (source.line as number) < 1)
            throw new Error('invalid source');
    }
    return run as unknown as Run;
}
export function readRunFile(directory: string, id: string, name: string): Buffer {
    if (!runIDPattern.test(id) || !['run.json', 'trace.jsonl', 'session.jsonl'].includes(name))
        throw new Error('invalid record path');
    ordinaryDirectory(directory);
    ordinaryDirectory(join(directory, id));
    return readOrdinaryFile(join(directory, id, name), name === 'session.jsonl' ? 16 * 1024 * 1024 : 64 * 1024 * 1024);
}
export function readStoredRun(directory: string, id: string): Run { return decodeHistory(readRunFile(directory, id, 'run.json'), id); }
export function loadHistory(directory: string) {
    const history = new Map<string, RunSummary>();
    let skipped = 0;
    if (existsSync(directory)) {
        // ponytail: scan once and retain summaries; add a persistent index if startup scans become slow.
        for (const id of readdirSync(directory)) {
            if (!runIDPattern.test(id))
                continue;
            try {
                history.set(id, summarizeRun(readStoredRun(directory, id)));
            }
            catch {
                skipped++;
            }
        }
    }
    return { history, skipped };
}

File: storage.go
package main

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"go/ast"
	"go/parser"
	"go/token"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"sort"
)

var runIDPattern = regexp.MustCompile("^[a-f0-9]{32}$")
var eventIDPattern = regexp.MustCompile("^e[0-9]+$")
var providerCodePattern = regexp.MustCompile("^[A-Za-z0-9_]{1,80}$")

type Source struct {
	Path string `json:"path"`
	Line int    `json:"line"`
	Code string `json:"code"`
}
type Event struct {
	ID          string         `json:"id"`
	Kind        string         `json:"kind"`
	Title       string         `json:"title"`
	Turn        int            `json:"turn"`
	T           float64        `json:"t"`
	D           float64        `json:"d"`
	Status      string         `json:"status"`
	Input       map[string]any `json:"input"`
	Output      map[string]any `json:"output"`
	Code        string         `json:"code"`
	Explanation string         `json:"explanation"`
}
type Run struct {
	Exercise         string            `json:"exercise,omitempty"`
	ParentRunID      string            `json:"parent_run_id,omitempty"`
	ConversationID   string            `json:"conversation_id,omitempty"`
	ConversationTurn int               `json:"conversation_turn,omitempty"`
	ID               string            `json:"id"`
	Task             string            `json:"task"`
	Model            string            `json:"model"`
	Workspace        string            `json:"workspace"`
	MaxRequests      int               `json:"max_requests"`
	Status           string            `json:"status"`
	TaskResult       string            `json:"task_result"`
	Answer           string            `json:"answer"`
	Error            map[string]any    `json:"error"`
	Events           []*Event          `json:"events"`
	CreatedAt        float64           `json:"created_at"`
	Duration         float64           `json:"duration"`
	ModelRequests    int               `json:"model_requests"`
	ToolCalls        int               `json:"tool_calls"`
	ToolErrors       int               `json:"tool_errors"`
	Source           map[string]Source `json:"source"`
	Engine           string            `json:"engine,omitempty"`
	BuildID          string            `json:"build_id,omitempty"`
}

type RunSummary struct {
	Exercise         string  `json:"exercise,omitempty"`
	ParentRunID      string  `json:"parent_run_id,omitempty"`
	ConversationID   string  `json:"conversation_id,omitempty"`
	ConversationTurn int     `json:"conversation_turn,omitempty"`
	ID               string  `json:"id"`
	Task             string  `json:"task"`
	Model            string  `json:"model"`
	Status           string  `json:"status"`
	CreatedAt        float64 `json:"created_at"`
}

func summarizeRun(run *Run) RunSummary {
	return RunSummary{Exercise: run.Exercise, ParentRunID: run.ParentRunID, ConversationID: run.ConversationID, ConversationTurn: run.ConversationTurn, ID: run.ID, Task: run.Task, Model: run.Model, Status: run.Status, CreatedAt: run.CreatedAt}
}

func sortedSummaries(history map[string]RunSummary) []RunSummary {
	items := make([]RunSummary, 0, len(history))
	for _, summary := range history {
		items = append(items, summary)
	}
	sort.Slice(items, func(i, j int) bool {
		if items[i].CreatedAt == items[j].CreatedAt {
			return items[i].ID < items[j].ID
		}
		return items[i].CreatedAt > items[j].CreatedAt
	})
	return items
}

func appendJSON(path string, value any) error {
	if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
		return err
	}
	file, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0600)
	if err != nil {
		return err
	}
	defer file.Close()
	encoder := json.NewEncoder(file)
	encoder.SetEscapeHTML(false)
	if err = encoder.Encode(value); err != nil {
		return err
	}
	return file.Sync()
}

func saveRun(path string, run *Run) error {
	file, err := os.CreateTemp(filepath.Dir(path), ".run-*.tmp")
	if err != nil {
		return err
	}
	defer os.Remove(file.Name())
	encoder := json.NewEncoder(file)
	encoder.SetEscapeHTML(false)
	encoder.SetIndent("", "  ")
	if err = encoder.Encode(run); err == nil {
		err = file.Sync()
	}
	closeErr := file.Close()
	if err != nil {
		return err
	}
	if closeErr != nil {
		return closeErr
	}
	return os.Rename(file.Name(), path)
}

func sourceRecords() (map[string]Source, string, error) {
	sources := map[string]Source{}
	hash := sha256.New()
	// This identifies the embedded source/asset bundle, not environment secrets or a binary checksum.
	if err := fs.WalkDir(assets, ".", func(path string, entry fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if entry.IsDir() {
			return nil
		}
		raw, err := assets.ReadFile(path)
		if err != nil {
			return err
		}
		hash.Write([]byte(path))
		hash.Write([]byte{0})
		hash.Write(raw)
		return nil
	}); err != nil {
		return nil, "", err
	}
	for _, definition := range []struct{ key, file, function string }{
		{"context", "context.go", "BuildTurnMessages"},
		{"coding_dispatch", "coding.go", "ExecuteCoding"}, {"command", "coding.go", "RunExerciseTests"}, {"write", "coding.go", "WriteExerciseFile"},
		{"loop", "agent.go", "RunLoop"}, {"dispatch", "tools.go", "ExecuteReadonly"},
		{"read", "tools.go", "ReadFile"}, {"list", "tools.go", "ListFiles"},
	} {
		raw, err := assets.ReadFile(definition.file)
		if err != nil {
			return nil, "", err
		}
		set := token.NewFileSet()
		tree, err := parser.ParseFile(set, definition.file, raw, 0)
		if err != nil {
			return nil, "", err
		}
		found := false
		for _, declaration := range tree.Decls {
			function, ok := declaration.(*ast.FuncDecl)
			if !ok || function.Name.Name != definition.function {
				continue
			}
			first, last := set.Position(function.Pos()), set.Position(function.End())
			sources[definition.key] = Source{Path: definition.file, Line: first.Line, Code: string(raw[first.Offset:last.Offset])}
			found = true
			break
		}
		if !found {
			return nil, "", errors.New("source function missing")
		}
	}
	return sources, hex.EncodeToString(hash.Sum(nil)), nil
}

func requiredObject(raw []byte, keys ...string) (map[string]json.RawMessage, error) {
	var fields map[string]json.RawMessage
	if json.Unmarshal(raw, &fields) != nil || fields == nil {
		return nil, errors.New("JSON object required")
	}
	for _, key := range keys {
		value, ok := fields[key]
		if !ok || bytes.Equal(bytes.TrimSpace(value), []byte("null")) {
			return nil, errors.New("required field missing")
		}
	}
	return fields, nil
}

func decodeHistory(raw []byte, id string) (*Run, error) {
	fields, err := requiredObject(raw, "id", "task", "model", "workspace", "status", "answer", "task_result",
		"created_at", "duration", "model_requests", "tool_calls", "tool_errors", "events", "source")
	if err != nil {
		return nil, err
	}
	var run Run
	if json.Unmarshal(raw, &run) != nil || run.ID != id || !runIDPattern.MatchString(run.ID) {
		return nil, errors.New("invalid run identity or fields")
	}
	if run.Exercise != "" && run.Exercise != codingExercise {
		return nil, errors.New("invalid exercise")
	}
	if run.Status != "completed" && run.Status != "failed" && run.Status != "budget_exhausted" {
		return nil, errors.New("run not finished")
	}
	if run.ParentRunID != "" && (!runIDPattern.MatchString(run.ParentRunID) || run.ParentRunID == run.ID || !runIDPattern.MatchString(run.ConversationID) || run.ConversationID == run.ID || run.ConversationTurn < 2) {
		return nil, errors.New("invalid conversation link")
	}
	if run.ParentRunID == "" && (run.ConversationID != "" || run.ConversationTurn > 1 || run.ConversationTurn < 0) {
		return nil, errors.New("invalid conversation root")
	}
	if run.CreatedAt < 0 || run.Duration < 0 || run.ModelRequests < 0 || run.ToolCalls < 0 || run.ToolErrors < 0 || len(run.Events) == 0 {
		return nil, errors.New("invalid history counters")
	}
	var eventRecords []json.RawMessage
	if json.Unmarshal(fields["events"], &eventRecords) != nil {
		return nil, errors.New("invalid events")
	}
	ids := map[string]bool{}
	var sourceRecords map[string]json.RawMessage
	if json.Unmarshal(fields["source"], &sourceRecords) != nil {
		return nil, errors.New("invalid sources")
	}
	for i, event := range run.Events {
		if _, err := requiredObject(eventRecords[i], "id", "kind", "title", "turn", "t", "d", "status", "input", "output", "code", "explanation"); err != nil {
			return nil, err
		}
		if event == nil || !eventIDPattern.MatchString(event.ID) || ids[event.ID] || event.T < 0 || event.D < 0 || event.Turn < 0 || event.Input == nil || event.Output == nil {
			return nil, errors.New("invalid event")
		}
		ids[event.ID] = true
		if event.Kind != "input" && event.Kind != "model" && event.Kind != "tool" && event.Kind != "control" {
			return nil, errors.New("invalid kind")
		}
		if event.Status != "succeeded" && event.Status != "failed" {
			return nil, errors.New("unfinished event")
		}
		if source, found := run.Source[event.Code]; !found || source.Line < 1 {
			return nil, errors.New("source missing")
		}
		if _, err := requiredObject(sourceRecords[event.Code], "path", "line", "code"); err != nil {
			return nil, err
		}
	}
	return &run, nil
}

func sortedRunIDs(runs map[string]*Run) []string {
	ids := make([]string, 0, len(runs))
	for id := range runs {
		ids = append(ids, id)
	}
	sort.Slice(ids, func(i, j int) bool {
		if runs[ids[i]].CreatedAt == runs[ids[j]].CreatedAt {
			return ids[i] < ids[j]
		}
		return runs[ids[i]].CreatedAt < runs[ids[j]].CreatedAt
	})
	return ids
}

func readStoredRun(directory, id string) (*Run, error) {
	raw, err := readRunFile(directory, id, "run.json")
	if err != nil {
		return nil, err
	}
	return decodeHistory(raw, id)
}

func readRunFile(directory, id, name string) ([]byte, error) {
	if !runIDPattern.MatchString(id) {
		return nil, errors.New("invalid run id")
	}
	if name != "run.json" && name != "trace.jsonl" && name != "session.jsonl" {
		return nil, errors.New("invalid record file")
	}
	folder := filepath.Join(directory, id)
	info, err := os.Lstat(folder)
	if err != nil {
		return nil, err
	}
	if !info.IsDir() || info.Mode()&os.ModeSymlink != 0 {
		return nil, errors.New("ordinary run directory required")
	}
	path := filepath.Join(folder, name)
	info, err = os.Lstat(path)
	if err != nil {
		return nil, err
	}
	if !info.Mode().IsRegular() || info.Mode()&os.ModeSymlink != 0 {
		return nil, errors.New("ordinary run file required")
	}
	if name == "session.jsonl" && info.Size() > 16<<20 {
		return nil, errors.New("对话记录已超过 16 MiB，请开始新任务")
	}
	return os.ReadFile(path)
}

func loadHistory(directory string) (map[string]RunSummary, int, error) {
	history := map[string]RunSummary{}
	skipped := 0
	entries, err := os.ReadDir(directory)
	if os.IsNotExist(err) {
		return history, 0, nil
	}
	if err != nil {
		return nil, 0, err
	}
	// ponytail: validate files once at startup and retain only summaries; add a persistent index if scanning becomes slow.
	for _, entry := range entries {
		run, readErr := readStoredRun(directory, entry.Name())
		if os.IsNotExist(readErr) {
			continue
		}
		if readErr != nil {
			skipped++
			continue
		}
		history[run.ID] = summarizeRun(run)
	}
	return history, skipped, nil
}
